blog/tanstack-router-intro/index.md

```diff
@@ -0,0 +1,461 @@
+
+![get search params](/tanstack-router-intro/get-search-params.jpg)
+
+We won't belabor the form elements and click handlers to sync and gather new values for these search parameters. Let's just assume we have some new values, and see how we set them. For this, we can use the `useNavigate` hook.
+
+```ts
+const navigate = useNavigate({ from: "/epics/$epicId/milestones/" });
+```
+
+We call it, and tell it where we're navigating _from_. Now we use the result and tell it where we want to _go_ (the same place we are), and are given a `search` function from which we return the new search params. Naturally, TypeScript will yell at us if we leave anything off. As a convenience, Router will pass this search function the current values, making it easy to just add / override something. So to page up, we can do
+
+```ts
+navigate({
+  to: ".",
+  search: prev => {
+    return { ...prev, page: prev.page + 1 };
+  },
+});
+```
+
+Naturally, there's also a `params` prop to this function, if you're browsing to a route with path parameters that you have to specify (or else TypeScript will yell at you, like always). We don't need an `$epicId` path param here, since there's already one on the route, and since we're going to the same place we already are (as indicated by the `from` value in useNavigate, and the `to: "."` value in navigate function) Router knows to just keep what's there, there.
+
+If we want to set a search value and tags, we could do
+
+```ts
+const newSearch = "Hello World";
+const tags = ["tag 1", "tag 2"];
+
+navigate({
+  to: ".",
+  search: prev => {
+    return { page: 1, search: newSearch, tags };
+  },
+});
+```
+
+which will make our URL look like this
+
+```
+/epics/1/milestones?page=1&search=Hello%20World&tags=%5B"tag%201"%2C"tag%202"%5D
+```
+
+Again, the search, and the array of strings were serialized for us.
+
+If we want to _link to_ a page with search params, we specify those search params on the Link tag
+
+```tsx
+<Link to="/epics/$epicId/milestones" params={{ epicId }} search={{ search: "", page: 1, tags: [] }}>
+  View milestones
+</Link>
+```
+
+And as always, TypeScript will yell at us if we leave anything off. Strong typing is a good thing.
+
+### Making our url prettier
+
+As we saw, currently, browsing to
+
+```
+http://localhost:5173/epics/1/milestones
+```
+
+will replace the url with this
+
+```
+http://localhost:5173/epics/1/milestones?page=1&search=&tags=%5B%5D
+```
+
+since we told Router that our page will always have a page, search, and tags value. If we want, we can make all of these values optional. In JavaScript (and TypeScript) a value does not exist if it holds the value `undefined`. So we could change our type to this
+
+```ts
+type SearchParams = {
+  page: number | undefined;
+  search: string | undefined;
+  tags: string[] | undefined;
+};
+```
+
+or this (which is the same thing)
+
+```ts
+type SearchParams = Partial<{
```

Comment:
in your case, all search params are optional.
however, if you have defaults, some params might never be undefined.
you then have a discrepancy between the return type of `validateSearch` and the input you'd like this to take

imagine this:


```tsx
type SearchParams = {
  page: number;
  search: string;
  tags: string[];
};

```ts
validateSearch(search: Record<string, unknown>): SearchParams {
  const page = Number(search.page ?? "1") ?? 1;
  const searchVal = (search.search as string) || "";
  const tags = Array.isArray(search.tags) ? search.tags : [];

  return {
    page,
    search: searchVal,
    tags,
  };
},
```

now Router knows that `page` will never be undefined. however, this will also require you to always specify `page` in search prop of  `navigate` / `<Link />`
if you don't want this, you can use the following

```tsx
import { SearchSchemaInput } from '@tanstack/react-router'

type SearchParamsInput = {
  page?: number;
  search?: string;
  tags?: string[]
}

type SearchParams = {
  page: number;
  search: string;
  tags: string[]
};

validateSearch(search: SearchParamsInput & SearchSchemaInput): SearchParams {
  const page = Number(search.page ?? "1") ?? 1;
  const searchVal = (search.search as string) || "";
  const tags = Array.isArray(search.tags) ? search.tags : [];

  return {
    page,
    search: searchVal || undefined,
    tags: tags.length ? tags : undefined,
  };
},
```

This allows do decouple the return type of `validateSearch` from the type that defines which properties Router requires you to specifiy.

full example at https://github.com/TanStack/router/blob/main/examples/react/basic-default-search-params/src/main.tsx